microcosm_pubsub/handlers/base.py

```diff
@@ -0,0 +1,160 @@
+"""
+Base pubsub handler
+
+"""
+from abc import ABCMeta
+from re import search
+
+from inflection import titleize
+from microcosm.errors import LockedGraphError, NotBoundError
+
+from microcosm_pubsub.constants import DEFAULT_RESOURCE_CACHE_TTL
+from microcosm_pubsub.conventions.lifecycle import LifecycleChange
+
+
+def resource_cache_whitelist_callable(media_type, uri):
+    """
+    Default resource cache whitelist callable implementation.
+
+    Only whitelists under the following conditions:
+    * verb is created
+    * resource is an event
+
+    """
+    if not media_type:
+        # Nb. likely to happen with mocks and tests.
+        return False
+
+    return all((
+        search(r"/[a-z]+_event/", uri),
+        search(r".{}.".format(LifecycleChange.Created), media_type),
+    ))
+
+
+class PubSubHandler(metaclass=ABCMeta):
+    """
+    Base handler for pubsub messages.
+
+    There are five expected outcomes for this handler:
+      - Raising an error (e.g. a bug)
+      - Skipping the handler.
+      - Handling the message.
+      - Ignoring the message.
+      - Raising a nack to reprocess at a later date (e.g. as a protection against race conditions)
+
+    The middle three cases are all handled here with the expectation that we produce *one* INFO-level
+    log per message processed (unless an error/nack is raised).
+
+    """
+    def __init__(
+        self,
+        graph,
+        retry_nack_timeout=1,
+        resource_nack_timeout=1,
+        resource_cache_enabled=True,
+        resource_cache_ttl=DEFAULT_RESOURCE_CACHE_TTL,
+        resource_cache_whitelist_callable=resource_cache_whitelist_callable,
+    ):
+        self.opaque = graph.opaque
+        self.retry_nack_timeout = retry_nack_timeout
+        self.resource_nack_timeout = resource_nack_timeout
+        self.resource_cache_ttl = resource_cache_ttl
+        self.resource_cache = self.get_resource_cache(graph) if resource_cache_enabled else None
+        self.resource_cache_whitelist_callable = resource_cache_whitelist_callable
+
+    @property
```

Comment:
We can probably just remove this if we're going all-new here.

Reply:
I'll move it to `URIHandler` instead of the base class.